Bottom: 7146352d6f140dbc64c9a05703abd00ca51415ca
Top:    11b1412b40650ea3164ef7471de1c35b8858f1d4
Author: Peter Clifton <peter@clifton-electronics.co.uk>
Date:   2015-01-01 17:10:54 +0000

draw.c: Opencode the mask layer object drawing functions

Calling DrawLayer() did not render using the correct graphics context / colors.
NB: We're still using the "wrong" GC, but we now avoid selecting inappropriate
colors which prevent (say) PS export working correctly for mask layer objects.


---

diff --git a/src/draw.c b/src/draw.c
index 26c0ff4..15c462d 100644
--- a/src/draw.c
+++ b/src/draw.c
@@ -871,6 +871,50 @@ DrawMaskBoardArea (int mask_type, const BoxType *drawn_area)
                                            drawn_area->X2, drawn_area->Y2);
 }
 
+static int
+mask_poly_callback (const BoxType * b, void *cl)
+{
+  struct poly_info *i = cl;
+  PolygonType *polygon = (PolygonType *)b;
+
+  gui->graphics->draw_pcb_polygon (Output.pmGC, polygon, i->drawn_area);
+  return 1;
+}
+
+static int
+mask_line_callback (const BoxType * b, void *cl)
+{
+  LineType *line = (LineType *)b;
+
+  gui->graphics->draw_pcb_line (Output.pmGC, line);
+  return 1;
+}
+
+static int
+mask_arc_callback (const BoxType * b, void *cl)
+{
+  ArcType *arc = (ArcType *)b;
+
+  gui->graphics->draw_pcb_arc (Output.pmGC, arc);
+  return 1;
+}
+
+static int
+mask_text_callback (const BoxType * b, void *cl)
+{
+  LayerType *layer = cl;
+  TextType *text = (TextType *)b;
+  int min_silk_line;
+
+  if (layer == &PCB->Data->SILKLAYER ||
+      layer == &PCB->Data->BACKSILKLAYER)
+    min_silk_line = PCB->minSlk;
+  else
+    min_silk_line = PCB->minWid;
+  gui->graphics->draw_pcb_text (Output.fgGC, text, min_silk_line);
+  return 1;
+}
+
 /* ---------------------------------------------------------------------------
  * draws solder mask layer - this will cover nearly everything
  */
@@ -878,6 +922,8 @@ void
 DrawMask (int side, const BoxType *screen)
 {
   int thin = TEST_FLAG(THINDRAWFLAG, PCB) || TEST_FLAG(THINDRAWPOLYFLAG, PCB);
+  LayerType *Layer = LAYER_PTR (side == TOP_SIDE ? top_soldermask_layer : bottom_soldermask_layer);
+  struct poly_info info;
 
   if (thin)
     gui->graphics->set_color (Output.pmGC, PCB->MaskColor);
@@ -887,7 +933,12 @@ DrawMask (int side, const BoxType *screen)
       gui->graphics->use_mask (HID_MASK_CLEAR);
     }
 
-  DrawLayer (LAYER_PTR (side == TOP_SIDE ? top_soldermask_layer : bottom_soldermask_layer), screen);
+  info.layer = Layer;
+  info.drawn_area = screen;
+  r_search (Layer->polygon_tree, screen, NULL, mask_poly_callback, &info);
+  r_search (Layer->line_tree,    screen, NULL, mask_line_callback, Layer);
+  r_search (Layer->arc_tree,     screen, NULL, mask_arc_callback,  Layer);
+  r_search (Layer->text_tree,    screen, NULL, mask_text_callback, Layer);
 
   r_search (PCB->Data->pin_tree, screen, NULL, clearPin_callback, NULL);
   r_search (PCB->Data->via_tree, screen, NULL, clearPin_callback, NULL);
